src/mpid/ch4/netmod/ucx/ucx_rma.h

```diff
@@ -375,9 +549,46 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_NM_mpi_rget(void *origin_addr,
                                                MPIR_Win * win,
                                                MPIDI_av_entry_t * addr, MPIR_Request ** request)
 {
-    return MPIDI_CH4U_mpi_rget(origin_addr, origin_count, origin_datatype,
-                               target_rank, target_disp, target_count, target_datatype, win,
-                               request);
+    int mpi_errno = MPI_SUCCESS;
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_NETMOD_UCX_RGET);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_NETMOD_UCX_RGET);
+
+    /* request based GET relies on UCX 1.4 function ucp_get_nb */
+#if defined(MPICH_UCX_AM_ONLY) || !defined(HAVE_UCP_GET_NB)
+    mpi_errno = MPIDI_CH4U_mpi_rget(origin_addr, origin_count, origin_datatype,
+                                    target_rank, target_disp, target_count, target_datatype, win,
+                                    request);
+#else
+    if (!MPIDI_UCX_is_reachable_target(target_rank, win)) {
+        mpi_errno = MPIDI_CH4U_mpi_rget(origin_addr, origin_count, origin_datatype,
+                                        target_rank, target_disp, target_count, target_datatype,
+                                        win, request);
+    } else {
+        MPIR_Request *sreq = NULL;
+
+        mpi_errno = MPIDI_UCX_do_get(origin_addr, origin_count, origin_datatype,
+                                     target_rank, target_disp, target_count, target_datatype,
+                                     win, addr, &sreq);
+        if (mpi_errno)
+            MPIR_ERR_POP(mpi_errno);
+
+        if (sreq == NULL) {
+            /* create a completed request for user if issuing is completed immediately. */
```

Comment:
Can we use the lightweight completed request here? `MPIDI_UCX_global.lw_send_req`. I should really unify that and the OFI version with the one recently added in MPIR for nonblocking collectives.

Reply:
I am trying to understand the lightweight request through code in `MPIDI_UCX_send`. It looks like, if I do not need to check the actual op completion (e.g., already completed), then I can use the `lightweight completed request`, right ?

Why do we disable it when `HAVE_DEBUGGER_SUPPORT` is set ? 

Moreover, do we want to create `MPIDI_UCX_global.lw_rma_req` with `MPIR_REQUEST_KIND__RMA` kind, for request-based RMA ?

Reply:
We disable it when debugger support is enabled so as not to corrupt the message queue interface. I don't think a separate RMA request is necessary. At least I am hoping not. I am trying to unify all the places we use the same technique in #3465, but for now you can try to use the ch4/ucx version.

Reply:
OK. The only difference in the request is the kind type. But I guess you might create a special type for all completed lightweight request. I will use ucx's `lw_send_req` in this PR.

Reply:
Yes, a special kind for this particular request is probably a good idea.

Reply:
1549617e09d80691fd5bd6421987f112cb38160c

Reply:
I see `MPIR_Process.lw_req` is with `MPIR_REQUEST_KIND__SEND`. Can we define a special kind for it since RMA may also use it ?

Reply:
We actually already use this for collective and RMA requests in the code today. l would also like to use it for send requests in the netmod, but that is currently breaking things in my testing. But your point is a good one. I'll add a patch to make it a more generic request type, which I think we still maintain for ch3.

Reply:
We may also change the completed request RMA in ofi and CH4R. These can go with a separate PR.